get_fedora_url.sh

```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+# This script is intended to be called by the Makefile, not by
+# humans.  This implies certain otherwise "odd" behaviors, such
+# as exiting with no std-output if there was an error.  It expects
+# to be called with three arguments:
+# 1. The type of url to retrieve, `image` or `checksum`.
+# 2. The architecture, `x86_64` or `aarch64`
+# 3. The Fedora release, 'rawhide' or a release number.
+
+set -eo pipefail
+
+URL_BASE="https://dl.fedoraproject.org/pub/fedora/linux/"
+CURL="curl --location --silent --fail"
+
+url_type="$1"
+arch_name="$2"
+fed_rel="$3"
+
+die() { echo "ERROR: ${1:-No error message provided}" > /dev/stderr; exit 1; }
+
+msg() { echo "${1:-No error message provided}" > /dev/stderr; }
+
+[[ "$#" -eq 3 ]] || \
+    die "Expecting exactly 3 arguments: <image|checksum> <x86_64|aarch64> <release #>"
+
+tmpfile=$(mktemp -p '' tmp.$(basename ${BASH_SOURCE[0]}).XXXX)
+
+unset output_url
+handle_exit() {
+    local msg_suffix="$arch_name Fedora $fed_rel release $url_type URL."
+    set +e
+    rm -f "$tmpfile" &> /dev/null
+    if [[ -n "$output_url" ]]; then
+        # msg "Found $msg_suffix"
+        echo -n "$output_url" > /dev/stdout
+        return 0
+    fi
+    die "Finding $msg_suffix"
+}
+trap handle_exit EXIT
+
+stage_tree="development"
+if [[ "$fed_rel" != "rawhide" ]] && $CURL -o /dev/null "${URL_BASE}/releases/$fed_rel"
+then
+    stage_tree="releases"
+fi
+
+cloud_download_url="${URL_BASE}/$stage_tree/$fed_rel/Cloud/$arch_name/images"
+
+$CURL -o "$tmpfile" "$cloud_download_url" || \
+    die "Retrieving directory listing."
+
+# msg "Searching Fedora $fed_rel $stage_tree tree for $url_type filename"
+
+# Will use a case-insensitive search
+file_regex_prefix="href=\"fedora.+$fed_rel.+$arch_name"
+
+if [[ "$url_type" == "image" ]]; then
+    file_regex="$file_regex_prefix.+qcow2\""
+elif [[ "$url_type" == "checksum" ]]; then
+    file_regex="$file_regex_prefix.+checksum\""
+else
+    die "Unknown/unsupported url type: '$url_type'."
+fi
+
+filename=$(egrep -i -m 1 -o "$file_regex" "$tmpfile" | cut -d '"' -f 2)
+output_url="$cloud_download_url/$filename"
```

Comment:
Some of this squicks me out a little: there's a lot that can fail with little visibility. I don't know what the chances are of that happening, but wdyt of
```bash
$CURL -o "$tmpfile" "$cloud_download_url" || \
    die "Could not fetch $cloud_download_url"

# msg "Searching Fedora $fed_rel $stage_tree tree for $url_type filename"

targets=$(sed -ne 's/^.*href=\"\(fedora[^\"]\+\)\".*$/\1/ip' <$tmpfile)
test -n "$targets" || die "Did not find any fedora targets in $cloud_download_url"

by_release=$(grep -w "$fed_rel" <<<"$targets" || true)
test -n "$by_release" || die "No '$fed_rel' targets among $targets"

by_arch=$(grep -w "$arch_name" <<<"$by_release" || true)
test -n "$by_arch" || die "No '$arch_name' targets among $by_release"

if [[ "$url_type" == "image" ]]; then
    extension=.qcow2
elif [[ "$url_type" == "checksum" ]]; then
    extension=-CHECKSUM
else
    die "Unknown/unsupported url type: '$url_type'."
fi
filename=$(fgrep -i -m 1 -- "$extension" <<<"$by_arch" || true)
test -n "$filename" || die "No '.$extension' targets among $by_arch"

# Found it!
echo "$cloud_download_url/$filename"
```

Reply:
Sure, I'm happy to copy-paste that, thanks for putting it together :grin:   The only thing that *squeaks* me out a little is the targets regex  I had to spend a whole 7-minutes groking it, then 13 more trying to un-cross my eyes :rofl:

Reply:
Sorry about that.
```bash
# extract all fedora link targets (e.g., fedora-foo-bar.qcow from <a href="fedora-foo-bar.qcow">fed... )
```

Reply:
Lol it's not you, just that regex tend quickly toward un-readability.  I did try it out and it works, so that's the most important thing :grin:   Working on integrating your suggestion and fixing up style to match the rest of the scripts here.  Still not sure about keeping/not the fancy exit handler or not.  It's a convenient place for outputting debugging details.  But I agree it's not so POLA.